DR2-1158 feat: Add ingest mapper lambda.

MancunianSam opened the pull request:
This needs https://github.com/nationalarchives/da-aws-clients/pull/29 to merge and deploy before the tests will pass but the code is ready for review.

* Reads the input from the step function step
* Gets title and description for department and series from discovery
* Parses the folder metadata csv
* Parses the asset metadata csv
* Parses the file metadata csv
* Parses the manifest file
* Converts these into Dynamo case classes
* Updates dynamo with the values
* Writes the state data for the next step function step.

techncl commented on src/test/scala/uk/gov/nationalarchives/MetadataServiceTest.scala:
```diff
@@ -0,0 +1,191 @@
+
+      result.size should equal(5 + seriesIdOpt.size)
+
+      def checkTableRows(ids: List[UUID], expectedSize: Int, expectedTable: DynamoTable) = {
```
This test is quite long, could this maybe be pulled out into a private method?

techncl commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -0,0 +1,100 @@
+      id: UUID,
+      parentPath: String,
+      name: String,
+      `type`: String,
```
Maybe `type` could be a sealed trait?

techncl replied:
Would editing the `toString` method (of each Type), to return a string representation of the class name, help?

techncl replied:
Also, could these be case objects instead of case classes and then you don't have to use the brackets?

MancunianSam replied:
I've changed them to case objects and I've added a `toString` method to the trait. We still need the custom formatter because Scanamo turns any Scala object into a dynamo object but it's cleaner with the case objects and `toString` now instead of calling `getClass.getSimpleName`

techncl commented on src/main/scala/uk/gov/nationalarchives/Lambda.scala:
```diff
@@ -0,0 +1,77 @@
+        input.s3Bucket,
+        input.s3Prefix,
+        folderMetadataIds ++ departmentSeriesIds,
+        folderMetadata.filter(fm => fm.title == fm.name).map(_.identifier),
```
Since you have the opposite of what you did on line 56, here, do you think, on line 56, it's worth using `partitionMap` instead of filter
```
    val (folderMetadataIdsWhereTitleAndNameSame, folderMetadataIdsWhereTitleAndNameNotSame) =
      folderMetadata.partitionMap {
        case fm if fm.title == fm.name => Left(fm.identifier)
        case fmTitleAndNameDoNotMatch => Right(fmTitleAndNameDoNotMatch.identifier)
      }

MancunianSam replied:
I don't know. It feels clearer as is at the expense of being slightly slower and I'd like to avoid constructing `Either` directly unless Scanamo makes me do it.

techncl replied:
Ok. Do you think `collect` might be better instead of a `filter` and `map`, so

` val folderMetadataIdsWhereTitleAndNameNotSame: List[UUID] = folderMetadata.collect{ fm if fm.title != fm.name => fm.identifier }`

and

`folderMetadata.collect{ case fm if fm.title != fm.name => fm.identifier }`?

MancunianSam replied:
Ok, makes sense, I've changed it.

techncl reviewed (commented):
Looks good, just some minor comments.

Also, these general ones about the tests:

1.  DiscoveryServiceTests - I noticed that there were no `getDepartmentAndSeriesRows` tests for when either the department, series or both are `None`, is there a possibility that either one is None?
2. LambdaTests - 
   1. I noticed that there was no test for what happens if the metadata service methods throw an error, the : `parseCsvWithHeaders`, `parseCsvWithoutHeaders` and `metadataToDynamoTables`
   2. Maybe a `verify` to see how maybe times `parseCsvWithHeaders` is called/not called in each test, to ensure thing are called when they are needed, not called when there's an error and in the correct order?

MancunianSam commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+    uses: nationalarchives/dr2-github-actions/.github/workflows/lambda_deploy.yml@main
+    with:
+      lambda-name: ${{ github.event.inputs.environment }}-dr2-ingest-mapper
+      deployment-package: notifications_lambda.zip
```
The template is wrong. I've put a PR to change it and changed this here.

MancunianSam commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+name: DR2 Deploy DR2 Ingest Mapper Lambda
```
Same with this, the duplicate DR2 is in the template. I've removed it from the template and from here.

MancunianSam commented on build.sbt:
```diff
@@ -0,0 +1,40 @@
+import uk.gov.nationalarchives.sbt.Log4j2MergePlugin.log4j2MergeStrategy
+
+ThisBuild / organization := "uk.gov.nationalarchives"
+ThisBuild / scalaVersion := "2.13.10"
```
I've bumped the version and changed the template.

MancunianSam commented on project/build.properties:
```diff
@@ -0,0 +1 @@
+sbt.version=1.9.1
```
I've bumped the version and changed the template.

techncl replied:
Did you up the version bump here?

MancunianSam replied:
No, sorry, did the template but not here.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -0,0 +1,99 @@
+  }
+}
+object MetadataService {
+  trait Metadata
```
1. This lambda isn't only for judgments.
2. Good idea, I've changed it to `sealed`
3. Good point, I've added those three to the trait.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -0,0 +1,99 @@
+    }
+  }
+}
+object MetadataService {
```
See comment above, this isn't only for judgments.

MancunianSam commented on src/test/scala/uk/gov/nationalarchives/LambdaTest.scala:
```diff
@@ -0,0 +1,310 @@
+  }
+  "handleRequest" should "return the correct values from the lambda" in {
+    val (folderIdentifier, assetIdentifier, _, _) = stubNetworkRequests()
+    val inJson =
```
Good point, I've moved it to a `private def`

MancunianSam commented on project/plugins.sbt:
```diff
@@ -0,0 +1,3 @@
+addSbtPlugin("com.eed3si9n" % "sbt-assembly" % "2.1.1")
+addSbtPlugin("org.scalameta" % "sbt-scalafmt" % "2.4.6")
```
I've bumped it here and in the template.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/DiscoveryService.scala:
```diff
@@ -0,0 +1,87 @@
+    }
+  }
+
+  private def callDiscoveryApi(citableReference: String): IO[DiscoveryAsset] = {
```
`getAssetFromDiscoveryApi` works.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/DiscoveryService.scala:
```diff
@@ -0,0 +1,87 @@
+      department <- input.department
+      series <- input.series
+    } yield for {
+      discoveryDepartment <- callDiscoveryApi(department)
```
`departmentDiscoveryAsset` is good.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/DiscoveryService.scala:
```diff
@@ -0,0 +1,87 @@
+      series <- input.series
+    } yield for {
+      discoveryDepartment <- callDiscoveryApi(department)
+      discoverySeries <- callDiscoveryApi(series)
```
`seriesDiscoveryAsset` is good.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -0,0 +1,100 @@
+
+  case class BagitManifest(checksum: String, filePath: String)
+
+  case class DepartmentSeries(department: DynamoTable, series: Option[DynamoTable])
```
`DepartmentAndSeriesTableData` make sense. I've changed it.

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/DiscoveryService.scala:
```diff
@@ -0,0 +1,87 @@
+object DiscoveryService {
+  case class DiscoveryScopeContent(description: String)
+  case class DiscoveryAsset(citableReference: String, scopeContent: DiscoveryScopeContent, title: String)
+  case class DiscoveryResponse(assets: List[DiscoveryAsset])
```
`DiscoveryCollectionAssetResponse` matches `DiscoveryCollectionAsset`

MancunianSam commented on src/main/scala/uk/gov/nationalarchives/DiscoveryService.scala:
```diff
@@ -0,0 +1,87 @@
+
+    for {
+      department <- input.department
+      series <- input.series
```
As I understand it, there will be no way for there to be a series but no department but someone will probably manage it. I've changed this so if there's no department but there is a series, you'll get the series and an unknown department.

MancunianSam commented on src/test/scala/uk/gov/nationalarchives/LambdaTest.scala:
```diff
@@ -0,0 +1,310 @@
+    val os = new ByteArrayOutputStream()
+    IngestMapperTest().handleRequest(is, os, null)
+    val dynamoRequestBodies = dynamoServer.getAllServeEvents.asScala.map(e => read[DynamoRequestBody](e.getRequest.getBodyAsString))
+    print(dynamoRequestBodies)
```
Oops

techncl commented on src/main/scala/uk/gov/nationalarchives/MetadataService.scala:
```diff
@@ -0,0 +1,99 @@
+object MetadataService {
+  trait Metadata
+
+  case class DynamoTable(
```
Sorry, for the first point, I think it was a copy and paste error; I think I was meant to say "Should this be called...".

MancunianSam commented:
There are existing tests for either department or series missing. I've added in another one "return an unknown department if the series and department are missing" for when they're both missing.

I've added a test into LambdaTest which returns an invalid CSV from S3 and so we get an error from the MetadataService. I don't want to have MetadataService mocked in Lambda test because it's supposed to be more of an end to end test so I don't think we can call verify.

There are existing tests in MetadataServiceTest to check for errors with csv files with an invalid format.

techncl reviewed (approved):
👍
